### PR TITLE
Start preparing for TypeJointMatrixINTEL switch

### PR DESCRIPTION
**Pull request by MrSidims**
The patch adds TypeJointMatrixINTELv2 which maps to new type OpCode
6184. Under new OpCode matrix type no longer has Layout parameter. The patch also moved 'scope' to optional matrix muladd instruction.

The changes are done only in the consumer part to prepare the switch and make E2E switch backward compatible by preparing consumers ahead of time.

Unfortunately there is no way to add a test foe this unless it's binary test, but it seems to be a bit unsafe to add this, so the patch was tested localy.

Spec change:
https://github.com/intel/llvm/pull/8175

### CODE REVIEW

**Comment by MrSidims** [comment]
@asudarsa @maksimsab @jgstarIntel please take a look

File: lib/SPIRV/libSPIRV/SPIRVType.h [hunk 1]
```diff
@@ -1085,11 +1090,25 @@ class SPIRVTypeJointMatrixINTEL : public SPIRVType {
...
   SPIRVValue *getColumns() const { return Args[1]; }
-  SPIRVValue *getLayout() const { return Args[2]; }
-  SPIRVValue *getScope() const { return Args[3]; }
-  SPIRVValue *getUse() const { return Args.size() > 4 ? Args[4] : nullptr; }
+  SPIRVValue *getLayout() const {
+    if (this->getOpCode() == internal::OpTypeJointMatrixINTEL)
+      return Args[2];
+    return nullptr;
+  }
+  SPIRVValue *getScope() const {
```
Comment (maksimsab) [inline lib/SPIRV/libSPIRV/SPIRVType.h:1098]: Could you please separate this methods by blank lines?

**Review by maksimsab — CHANGES_REQUESTED** [review]
I see in the patch description that `scope is moved to optional matrix muladd instruction`.

Could you please clarify where is it done? Unfortunately, not obvious for me :(

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 2]
```diff
@@ -3329,10 +3330,10 @@ class SPIRVJointMatrixINTELInst : public SPIRVJointMatrixINTELInstBase {
...
 _SPIRV_OP(JointMatrixLoad, true, 6, true)
 _SPIRV_OP(JointMatrixStore, false, 5, true)
-_SPIRV_OP(JointMatrixMad, true, 7)
-_SPIRV_OP(JointMatrixSUMad, true, 7)
-_SPIRV_OP(JointMatrixUSMad, true, 7)
-_SPIRV_OP(JointMatrixUUMad, true, 7)
+_SPIRV_OP(JointMatrixMad, true, 6, true)
+_SPIRV_OP(JointMatrixSUMad, true, 6, true)
+_SPIRV_OP(JointMatrixUSMad, true, 6, true)
+_SPIRV_OP(JointMatrixUUMad, true, 6, true)
```
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:3336]: @maksimsab move of scope parameter happens here (see the second 'true').

Note, that 'scope' according to old spec is mandatory parameter and per new spec is removed at all. So to balance it here making both representations translatable we have to add is as optional for a while.

File: lib/SPIRV/libSPIRV/SPIRVType.cpp [hunk 3]
```diff
@@ -279,13 +280,19 @@ void SPIRVTypeForwardPointer::decode(std::istream &I) {
...
+    SPIRVModule *M, SPIRVId TheId, Op OC, SPIRVType *CompType,
     std::vector<SPIRVValue *> Args)
     : SPIRVType(M, FixedWC + Args.size(), OC, TheId), CompType(CompType),
       Args(Args) {}
 
+SPIRVTypeJointMatrixINTEL::SPIRVTypeJointMatrixINTEL(
+    SPIRVModule *M, SPIRVId TheId, SPIRVType *CompType,
+    std::vector<SPIRVValue *> Args)
+    : SPIRVType(M, FixedWC + Args.size(), internal::OpTypeJointMatrixINTEL,
+                TheId), CompType(CompType), Args(Args) {}
```
[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVType.cpp]: Thanks! We probably should also scan other places in the translator to embrace this optimization

**Review by maksimsab — APPROVED** [review]
LGTM! Thank you

**Pull request merged**